processor/tile_grpc.go

```diff
@@ -43,12 +67,62 @@ func (gi *GeoRasterGRPC) Run() {
 		log.Fatalf("gRPC connection problem: %v", err)
 	}
 	defer conn.Close()
+
+	// We have to do this because there is no way to figure out
+	// the data type of the raster but returned from the grpc worker
+	// We need the data type to dertermine the byte size for memory
+	// bound calcuations
+	g0 := grans[0]
+	r0, err := getRpcRaster(gi.Context, g0, conn)
+	if err != nil {
+		gi.Error <- err
+		r0 = &pb.Result{Raster: &pb.Raster{Data: make([]uint8, g0.Width*g0.Height), RasterType: "Byte", NoData: -1.}}
+		gi.Out <- &FlexRaster{ConfigPayLoad: g0.ConfigPayLoad, Data: r0.Raster.Data, Height: g0.Height, Width: g0.Width, OffX: g0.OffX, OffY: g0.OffY, Type: r0.Raster.RasterType, NoData: r0.Raster.NoData, NameSpace: g0.NameSpace, TimeStamp: g0.TimeStamp, Polygon: g0.Polygon}
+		return
+	}
+
+	dataSize, err := getDataSize(r0.Raster.RasterType)
+	if err != nil {
+		gi.Error <- err
+		return
+	}
+	requestedSize := imageSize * dataSize * len(grans)
+
+	meminfo := utils.MemInfo{}
+	err = meminfo.Update()
+
+	if err == nil {
+		freeMem := int(meminfo.Available())
+		if freeMem-requestedSize <= ReservedMemorySize {
+			log.Printf("Out of memory, freeMem:%v, requested:%v", freeMem, requestedSize)
+			gi.Error <- fmt.Errorf("Server resources exhausted")
+			return
+		}
+
+	} else {
+		// If we have error obtaining meminfo,
+		// we assume that we have enough memory.
+		// This can happen if the OS doesn't support
+		// /proc/meminfo
+		log.Printf("meminfo error: %v", err)
+	}
+
+	// We first send out the raster we used to figure out data type
+	gi.Out <- &FlexRaster{ConfigPayLoad: g0.ConfigPayLoad, Data: r0.Raster.Data, Height: g0.Height, Width: g0.Width, OffX: g0.OffX, OffY: g0.OffY, Type: r0.Raster.RasterType, NoData: r0.Raster.NoData, NameSpace: g0.NameSpace, TimeStamp: g0.TimeStamp, Polygon: g0.Polygon}
+
+	timeoutCtx, cancel := context.WithTimeout(gi.Context, time.Duration(g0.Timeout)*time.Second)
+	defer cancel()
 	cLimiter := NewConcLimiter(16)
-	for gran := range gi.In {
+	for i := 1; i < len(grans); i++ {
```

Comment:
Why not use `range` here?

Reply:
This is because we want to skip the first elements (i.e. i starts from 1 instead 0). The first element is already sent to the grpc worker in order to determine the data type of the raster.

Reply:
OK, that's fine then.